tools/lsp/preview/properties.rs

```diff
@@ -7,31 +7,62 @@ use crate::wasm_prelude::*;
 use crate::common::{self, Result, SourceFileVersion};
 use crate::util;
 
-use i_slint_compiler::diagnostics::Spanned;
+use i_slint_compiler::diagnostics::{SourceFile, Spanned};
 use i_slint_compiler::langtype::{ElementType, Type};
 use i_slint_compiler::object_tree::{Element, PropertyDeclaration, PropertyVisibility};
-use i_slint_compiler::parser::{syntax_nodes, Language, SyntaxKind, TextRange, TextSize};
+use i_slint_compiler::parser::{
+    syntax_nodes, Language, SyntaxKind, SyntaxNode, TextRange, TextSize,
+};
 
 use lsp_types::Url;
 
 use std::collections::HashSet;
 use std::path::PathBuf;
 
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Debug)]
+pub enum CodeBlockOrExpression {
+    CodeBlock(syntax_nodes::CodeBlock),
+    Expression(syntax_nodes::Expression),
+}
+
+impl CodeBlockOrExpression {
+    pub fn new_from(source_file: &SourceFile, node: &rowan::SyntaxNode<Language>) -> Option<Self> {
```

Comment:
would be nice to not use the rowan stuff directly here and try to use just the type exposed by the Slint's parser
It means you can't use `parent_ancestors` in `find_code_block_or_expression` but just call .parent() in a while loop instead. (lookup_current_element_type or find_element_id_for_highlight has an example of such while loop)

Reply:
Wouldn't I need to change that into an Expression later anyway?

Hmmm... might be a good idea indeed, assuming I am going to ignore `CodeBlock`s for the most part anyway. I can create a proper `Expression` node when I actually need that. I need to think about that a bit more.

Reply:
The type directly exposed by the parser is `BindingExpression`, which is either the `CodeBlock` or `Expression` I want, but surrounded by extra whitespace and a `;` Token. That makes it hard to get a useful `text_range()` out of the `BindingExpression`.

Not having two constructors for this type felt simpler, but I'll take a look agian.

Reply:
I mean to use `parser::SyntaxNode` instead of `rowan::SyntaxNode<Language>` (and SourceFile)